### ISSUE TITLE
forall causes a runtime error

### ISSUE THREAD

**Issue by rohit507** [dreal/dreal4#86]
```python
from dreal.symbolic import *
from dreal.api import CheckSatisfiability, Minimize

def test_dreal_interface_find_othrogonal():
    """
    Just an empty test to make pytest force a parse check.
    """
    t = Variable('theta',Variable.Real)
    a = Variable('a',Variable.Int)
    b = Variable('b',Variable.Int)
    c = Variable('c',Variable.Int)
    d = Variable('d',Variable.Int)
    fun = forall([t], logical_imply(logical_and(t < 3,t > -3),
                                      (((sin(t)
                                         * if_then_else(b == True, 1 , -1)
                                         * if_then_else(a == True,sin(t),cos(t)))
                                        + (cos(t)
                                           * if_then_else(d == True, 1 , -1)
                                           * if_then_else(c == True,sin(t),cos(t))))
            == 0)))
    result = CheckSatisfiability(fun, 0.01)
    print(result)

test_dreal_interface_find_othrogonal()
```

The above equation, which should be true for (a = False, b = True, c = True, d = False), instead causes an error `RuntimeError: Not yet supported.` 

There's also a similar error when I initialize any of the integer variables with `Variable.Bool` but things seem to work fine if I initialize them as integers and decode them back into floats myself.

What unimplemented feature is causing this error? It might be something I can work around, but I'm having trouble finding it.

**Comment by soonho-tri** [comment]
I guess the issue is that I didn't consider the case where we have Boolean variables inside of the universal quantifiers which we have in your example (via `if_then_else`).

I'll take a closer look later today.

**Comment by soonho-tri** [comment]
BTW,

```python
    if_then_else(b == True, 1 , -1)
```

`b == True` doesn't sound correct given the fact `b : Int`.

**Comment by rohit507** [comment]
It is not, but without it I get an error saying that Variables are not allowed in that position of `if_then_else`. In practice `True == 1` and `False == 0` but I've written it as above for readability.

**Comment by soonho-tri** [comment]
There are two issues here:

1) If-then-else eliminator does not support `forall` formulas yet. https://github.com/dreal/dreal4/blob/5e7c80ab49ac976718da020acc959480640a31d7/dreal/util/if_then_else_eliminator.cc#L336-L357

2) There is an ambiguity in `formula == formula` (i.e. `iff`) and `expr == expr` (i.e. `eq`).

I'll address them later this week (probably after Wed). Thanks for the patience.

**Comment by rohit507** [comment]
So, I've got a more representative example of what I'm working with dreal to do. Or at least one possible version thereof. (Here's the gist)[https://gist.github.com/rohit507/7c6e9abb7d90b2418c5310a5e05fb27d] with a toy problem. 

It produces an identical runtime error as the above, but likely for different reasons. 

Long story short, I can take coordinate transforms and other relationships between mechanical systems and flatten them into statements within DReal's theory (at least, I think they are). I want to be able to find a set of parameters for a mechanical system that which can be actuated to meet some requirements over its entire operating range. 

The gist has a simple example where I want to find the size and origin of an arm so that it is possible for it to reach all the points within a circle. The example is in 2D, though the tool I'm working on is for problems in 3D. 

TL;DR: Grad school is fun.

**Comment by soonho-tri** [comment]
Hi @rohit507 , as Sean mentioned we only support one alternation of quantifiers (exists-forall). If you mean nested "universal" quantifiers, then it's easy to collapse them into a single quantifier with a block of variables.

I'll work on this issue as my time allows.

**Comment by soonho-tri** [comment]
FTR, ITE elimination introduces a fresh (existentially-quantified) variable. For example,
`ITE(f, e₁, e₂) > 0` will be translated into `∃v. (v > 0) ∧ (f => (v = e₁)) ∧ (¬f => (v = e₂))`.

In case of `∃∀`-problems, this is not desirable since we end up with ∃∀∃-formulas. I think a better approach is to defer the elimination inside of the `ContractorForall` where we only handle ground formulas.

**Comment by soonho-tri** [comment]
Got it. I'll check.
